File: test_calculator.js

// Mock of MATERIAL_RULES from products_data.js
const MATERIAL_RULES = {
    "Drop Flat/Tenzado 1 hilo (Bobina 1km)": { "unit": "Bobina", "qty_per_unit": 1000.0, "rounding": "ESTRICTO", "logic": "Fibra" },
    "CABLE DE FIBRA PTICA 12 HILOS ADSS SPAN 100": { "unit": "Bobina", "qty_per_unit": 4000.0, "rounding": "LIBRE", "logic": "Fibra" },
    "Hebilla 1/2": { "unit": "Unidad", "qty_per_unit": 100.0, "rounding": "LIBRE", "logic": "Manual" }
};

// Copy of MaterialCalculator class from script.js
class MaterialCalculator {
    static calculate(itemName, quantityNeeded) {
        // 1. Obtener regla del catálogo
        const rule = (typeof MATERIAL_RULES !== 'undefined') ? MATERIAL_RULES[itemName] : null;

        if (!rule) {
            return {
                originalItem: itemName,
                originalQty: quantityNeeded,
                finalItem: itemName,
                finalQty: quantityNeeded,
                unit: 'u',
                note: 'Cálculo directo (Sin regla definida)'
            };
        }

        const qtyPerUnit = rule.qty_per_unit || 1;
        const unitName = rule.unit || 'u';
        const rounding = rule.rounding || 'LIBRE'; // 'ESTRICTO' o 'LIBRE'

        let finalQty = 0;
        let note = '';

        // 2. Lógica por Tipo de Unidad
        if (unitName.toLowerCase().includes('bobina') || unitName.toLowerCase().includes('rollo')) {
            // CASO: CABLE / BOBINAS
            const packsNeeded = quantityNeeded / qtyPerUnit;

            if (rounding === 'ESTRICTO') {
                // Forzar entero superior
                const packsInt = Math.ceil(packsNeeded);
                finalQty = packsInt;
                note = `Redondeado a ${packsInt} ${unitName}(s) de ${qtyPerUnit}m (Req: ${quantityNeeded}m)`;
            } else {
                // LIBRE: 2 decimales
                finalQty = parseFloat(packsNeeded.toFixed(2));
                note = `Equivalente a ${finalQty} ${unitName}(s) (Req: ${quantityNeeded}m)`;
            }

        } else {
            // CASO: UNIDADES DISCRETAS
            if (rounding === 'ESTRICTO' || (qtyPerUnit > 1)) {
                const packsNeeded = quantityNeeded / qtyPerUnit;
                finalQty = Math.ceil(packsNeeded);
                note = `Pack de ${qtyPerUnit} u. (Req: ${quantityNeeded})`;
            } else {
                finalQty = Math.ceil(quantityNeeded);
                note = '';
            }
        }

        return {
            originalItem: itemName,
            originalQty: quantityNeeded,
            finalItem: itemName,
            finalQty: finalQty,
            unit: unitName,
            note: note,
            ruleApplied: rule
        };
    }
}

// TEST CASES
console.log("--- TEST INICIO ---");

// Test 1: Drop Cable (Strict Rounding)
const t1 = MaterialCalculator.calculate("Drop Flat/Tenzado 1 hilo (Bobina 1km)", 1500);
console.log("Test 1 (1500m Drop -> Bobinas):", t1.finalQty, t1.unit, "| Expected: 2 Bobina");
if (t1.finalQty === 2 && t1.unit === "Bobina") console.log("✅ PASS"); else console.log("❌ FAIL");

// Test 2: ADSS Cable (Free Rounding)
const t2 = MaterialCalculator.calculate("CABLE DE FIBRA PTICA 12 HILOS ADSS SPAN 100", 6000);
// 6000 / 4000 = 1.5
console.log("Test 2 (6000m ADSS -> Bobinas):", t2.finalQty, t2.unit, "| Expected: 1.5 Bobina");
if (t2.finalQty === 1.5) console.log("✅ PASS"); else console.log("❌ FAIL");

// Test 3: Hebillas (Free Rounding but QtyPerUnit > 1 implies pack logic usually, strictly speaking 'Unit' with '100' per unit)
// Rules says: rounding LIBRE.
// Logic says: if qtyPerUnit > 1, use Pack logic?
// script.js logoc: if (rounding === 'ESTRICTO' || (qtyPerUnit > 1)) -> Math.ceil.
// So 150 hebillas / 100 = 1.5 packs -> ceil -> 2 packs.
const t3 = MaterialCalculator.calculate("Hebilla 1/2", 150);
console.log("Test 3 (150 Hebillas -> Packs of 100):", t3.finalQty, t3.unit, "| Expected: 2 Unidad");
if (t3.finalQty === 2) console.log("✅ PASS"); else console.log("❌ FAIL");

console.log("--- TEST FIN ---");

File: test_calculator.py
import math

# Mock of MATERIAL_RULES
MATERIAL_RULES = {
    "Drop Flat/Tenzado 1 hilo (Bobina 1km)": { "unit": "Bobina", "qty_per_unit": 1000.0, "rounding": "ESTRICTO", "logic": "Fibra" },
    "CABLE DE FIBRA PTICA 12 HILOS ADSS SPAN 100": { "unit": "Bobina", "qty_per_unit": 4000.0, "rounding": "LIBRE", "logic": "Fibra" },
    "Hebilla 1/2": { "unit": "Unidad", "qty_per_unit": 100.0, "rounding": "LIBRE", "logic": "Manual" }
}

class MaterialCalculator:
    @staticmethod
    def calculate(item_name, quantity_needed):
        rule = MATERIAL_RULES.get(item_name)

        if not rule:
            return {
                "finalQty": quantity_needed,
                "unit": 'u',
                "note": 'Cálculo directo'
            }

        qty_per_unit = rule.get("qty_per_unit", 1)
        unit_name = rule.get("unit", 'u')
        rounding = rule.get("rounding", 'LIBRE')

        final_qty = 0
        note = ''

        unit_lower = unit_name.lower()
        if 'bobina' in unit_lower or 'rollo' in unit_lower:
            packs_needed = quantity_needed / qty_per_unit
            
            if rounding == 'ESTRICTO':
                final_qty = math.ceil(packs_needed)
                note = f"Redondeado a {final_qty} {unit_name}(s)"
            else:
                # LIBRE: 2 decimals, simulates JS behavior
                final_qty = round(packs_needed + 0.0000001, 2) # small epsilon for float stability if needed
                # Actually JS toFixed(2) rounds half up/down depending on implementation but mostly standard.
                # Python round is banker's rounding, but for simple division it should be close enough.
                # Let's just use simple formatting or math.
                final_qty = float(f"{packs_needed:.2f}")
                note = f"Equivalente a {final_qty} {unit_name}(s)"
        else:
            # Discrete units
            if rounding == 'ESTRICTO' or qty_per_unit > 1:
                packs_needed = quantity_needed / qty_per_unit
                final_qty = math.ceil(packs_needed)
                note = f"Pack de {qty_per_unit} u."
            else:
                final_qty = math.ceil(quantity_needed)
                note = ''
        
        return {
            "finalQty": final_qty,
            "unit": unit_name
        }

print("--- TEST INICIO ---")

# Test 1
res1 = MaterialCalculator.calculate("Drop Flat/Tenzado 1 hilo (Bobina 1km)", 1500)
print(f"Test 1 (1500m Drop -> Bobinas): {res1['finalQty']} {res1['unit']} | Expected: 2 Bobina")

# Test 2
res2 = MaterialCalculator.calculate("CABLE DE FIBRA PTICA 12 HILOS ADSS SPAN 100", 6000)
print(f"Test 2 (6000m ADSS -> Bobinas): {res2['finalQty']} {res2['unit']} | Expected: 1.5 Bobina")

# Test 3
res3 = MaterialCalculator.calculate("Hebilla 1/2", 150)
print(f"Test 3 (150 Hebillas -> Packs 100): {res3['finalQty']} {res3['unit']} | Expected: 2 Unidad")

print("--- TEST FIN ---")
